File: Ywy/JavaScript/level2/CheckDistance.js
export default class CheckDistance {
    solution = (places) => {
        const answer = [];
        places.forEach(place => {
            answer.push(this.roomSearch(place));
        });
        return answer;
    };

    roomSearch = (place) => {
        let rooms = undefined;
        let row = 0;
        let result = 1;
        const range = place.length;
        do {
            rooms = place[row];
            if (rooms.indexOf("P") === -1) {
                continue;
            }
            result = this.search(place, [...rooms], row);
            if (result == 0) {
                break;
            }
        } while (++row < range);
        return result;
    };

    search = (place, rooms, row) => {
        let range = rooms.length;
        let col = 0;
        let room = undefined;
        do {
            room = rooms[col];
            if (room === 'P' && this.move(place, rooms, row, col) === 0) {
                return 0;
            }
        } while (++col < range);
        return 1;
    };

    move = (place, rooms, row, col) => {
        const moves = [(place[row - 1] || [])[col], rooms[col + 1], (place[row + 1] || [])[col], rooms[col - 1]];
        let move = 0;
        let index = 0;
        let range = moves.length;
        let flag = false;
        do {
            move = moves[index];
            if (move === undefined) {
                continue;
            }
            if (move === 'P') {
                return 0;
            } else if (move === 'X') {
                continue;
            }
            switch (index) {
                case 0:
                    flag = (place[row - 1] || [])[col + 1] === 'P' || (place[row - 2] || [])[col] === 'P';
                    break;
                case 1:
                    flag = (place[row + 1] || [])[col + 1] === 'P' || rooms[col + 2] === 'P';
                    break;
                case 2:
                    flag = (place[row + 1] || [])[col - 1] === 'P' || (place[row + 2] || [])[col] === 'P';
                    break;
                case 3:
                    flag = (place[row - 1] || [])[col - 1] === 'P' || rooms[col - 2] === 'P';
                    break;
            }
            if (flag) {
                return 0;
            }
        } while (++index < range);
        return 1;
    };
}

File: Ywy/JAVA/src/level2/CheckDistance.java
package level2;

import java.util.HashMap;
import java.util.Map;

public class CheckDistance {
	public int[] solution(String[][] places) {
		final int[] answer = new int[5];
		int index = 0;
		int range = places.length;
		while(true) {
			answer[index] = check(places[index++]);
			if (index == range) {
				return answer;
			}
			
		}
		
	}

	private int check(final String[] place) {
		Map<String, Boolean> partitionPos = new HashMap<>();
		int row = 0;
		int range = place.length;
		int result = 0;
		String room = null;
		do {
			room = place[row];
			if (room.indexOf("P") == -1) {
				if (room.indexOf("X") > -1) {
					putPartition(room, row, partitionPos);
				}
				continue;
			}
			
			result = search(place, room, row, partitionPos);
			if (result == 0) {
				return 0;
			}
		} while(++row < range);
			
		return 1;
	}
	
	private void putPartition(String room, int row, Map<String, Boolean> partitionPos) {
		int col = 0;
		int range = room.length();
		char at = 0;
		String key = null;
		do { 
			at = room.charAt(col);
			if (at == 'X') {
				key = getKey(row, col);
				if (key == null) {
					partitionPos.put(key, true);
				}
			}
		} while(++col < range);
	}

	private int search(String[] rooms, String room, int row, Map<String, Boolean> partitionPos) {
		char[] el = room.toCharArray();
		int col = 0;
		int range = el.length;
		char current = 0_0;
		int temp = 0;
		while(col < range) {
			current = el[col];
			if (current == 'X') {
				partitionPos.put(getKey(row, col), true);
				col++;
				continue;
			} 
			
			if (current != 'P') {
				col++;
				continue;
			}
			
			System.out.print("시작:");
			System.out.println(getKey(row, col) + " = " + current);
			
			if (col > 0) {

				temp = horizontalSearch(rooms, el, row, col, false, partitionPos);
				if (temp == 0) {
					return 0;
				}
			}
			
			if (col < range - 1) {
				temp = horizontalSearch(rooms, el, row, col, true, partitionPos);
				if (temp == 0) {
					return 0;
				}
			}
			
			if (row > 0) {
				temp = verticalSearch(rooms, row, col, true, partitionPos, 0);
				if (temp == 0) {
					return 0;
				}
			} 
			
			if (row < 5) {
				temp = verticalSearch(rooms, row, col, false, partitionPos, 0);
				if (temp == 0) {
					return 0;
				}
			}
			
			col++;
		}
		return 1;
	}
	
	private int verticalSearch(String[] rooms, int row, int col, boolean flag, Map<String, Boolean> partitionPos, int count) {
		char temp = 0;
		while(true) {
			if (count == 2 || (flag ? row - 1 < 0 : row + 1 > 4)) {
				return 1;
			}

			count++;
			temp = rooms[flag ? --row : ++row].charAt(col);
			System.out.println((!flag ? "▽" : "▲") + getKey(row, col) + " = " + temp);
			if(count == 2 && temp == 'P') {
				return partitionPos.get(getKey((flag? row + 1 : row - 1), col)) != null ? 1: 0;
			} else  if (temp == 'P') {
				return 0;
			} 
			
			if (temp == 'X') {
				partitionPos.put(getKey(row, col), true);
			}
		}
	}
	
	private int horizontalSearch(String[] rooms, char[] el, int row, int col, boolean flag, Map<String, Boolean> partitionPos) {
		char temp = 0;
		int count = 0;
		while(true) {
			if (count == 2 || (flag ? col + 1 > 4 : col - 1 < 0)) {
				return 1;
			}

			count++;
			temp = el[flag? ++col : --col];
			System.out.println((flag ? "▷" : "◀") + getKey(row, col) + " = " + temp);
			if(count == 2 && temp == 'P') {
				return partitionPos.get(getKey(row, (flag? col - 1 : col + 1))) != null ? 1 : 0;
			}
			
			if (temp == 'P') {
				return 0;
			} 
			
			if (temp == 'X') {
				partitionPos.put(getKey(row, col), true);
			}
			
			if (count == 1 && row > 0) {
				temp = (char) verticalSearch(rooms, row, col, true, partitionPos, 1);
				if (temp == 0) {
					return 0;
				}
			} 
			
			if (count == 1 && row < 5) {
				temp = (char) verticalSearch(rooms, row, col, false, partitionPos, 1);
				if (temp == 0) {
					return 0;
				}
			}
		}
	}

	private String getKey(int row, int col) {
		return String.format("%d_%d", row, col);
	}
}
